docs/contributors/code/e2e/overusing-snapshots.md

```diff
@@ -0,0 +1,112 @@
+await page.keyboard.type( '1' );
+await page.keyboard.press( 'Enter' );
+await page.keyboard.press( 'Enter' );
+
+expect( await editor.getEditedPostContent() ).toMatchSnapshot();
+
+await page.keyboard.press( 'Backspace' );
+await page.keyboard.type( '2' );
+
+expect( await editor.getEditedPostContent() ).toMatchSnapshot();
+```
+
+This is borrowed from the real code in gutenberg, with the test title and the comments removed and refactored into Playwright. Ideally, E2E tests should be self-documented and readable to end users; in the end, they are trying to resemble how end users interact with the app. However, there are a couple of red flags in the code.
+
+## Problems with snapshot testing
+
+Popularized by Jest, [snapshot testing](https://jestjs.io/docs/snapshot-testing) is a great tool to help test our app _when it makes sense_. However, probably because it's so powerful, it's often overused by developers. There are already multiple [articles](https://kentcdodds.com/blog/effective-snapshot-testing) about this. In this particular case, snapshot testing fails to reflect the developer's intention. It's not clear what the assertions are about without looking into other information. This makes the code harder to understand and creates a mental overhead for all the other readers other than the one who wrote it. As readers, we have to jump around the code to fully understand them. The added complexity of the code discourages contributors from changing the test to fit their needs. It could sometimes even confuse the authors and make them accidentally [commit the wrong snapshots](https://github.com/WordPress/gutenberg/pull/42780#discussion_r949865612).
+
+Here's the same test with the test title and comments. Now you know what these assertions are actually about.
+
+```js
+it( 'can be split at the end', async () => {
+	// ...
+
+	// Expect empty paragraph outside quote block.
+	expect( await getEditedPostContent() ).toMatchSnapshot();
+
+	// ...
+
+	// Expect the paragraph to be merged into the quote block.
+	expect( await getEditedPostContent() ).toMatchSnapshot();
+} );
+```
+
+The developer's intention is a bit more readable, but it still feels disconnected from the test. You might be tempted to try [inline snapshots](https://jestjs.io/docs/snapshot-testing#inline-snapshots), which does solve the issue of having to jump around files, but it's still not self-documented nor explicit. We can do better.
+
+## The solution
+
+Instead of writing the assertions in comments, we can try directly writing them out explicitly. With the help of `editor.getBlocks`, we can rewrite them into simpler and atomic assertions.
+
+```js
+// ...
+
+{
+	// Expect empty paragraph outside quote block.
+	const blocks = await editor.getBlocks();
+	expect( blocks ).toHaveLength( 2 );
+	expect( blocks[ 0 ] ).toMatchObject( {
+		name: 'core/quote',
+		innerBlocks: [
+			{
+				name: 'core/paragraph',
+				attributes: { content: '1' },
+			},
+		],
+	} );
```

Comment:
Should we also mention serialized content assertions as well?

Getting the serialized expected results are easier by switching to the code editor.

```js
await expect.poll( editor.getEditedPostContent )
			.toBe( `<!-- wp:paragraph -->
<p>1</p>
<!-- /wp:paragraph -->` );
```

Reply:
Good idea! I think we should be using `editor.getBlocks` in these cases too, for the same reason as snapshot testing.

Reply:
I think I prefer to use serialized inline assertions for simple text blocks. Let's document both use variations, and we can choose depending on the test case. Unless you think this would make standardizing the tests harder.


P.S. I might change my mind when I use `editor.geBlocks` and `toMatchObject` more often.

Reply:
I added a section about "snapshot variants". LMK what you think!

Reply:
Oops, didn't see your last comment.

I think I'd prefer `getBlocks` wherever possible. But it's good because this is what this RFC is for 😆 . Do you have a concrete example of where you think serialized blocks make more sense?

Reply:
Thank you, the update looks great.

>Do you have a concrete example of where you think serialized blocks make more sense?

I guess it's a matter of personal preference or I got used to this flow. But I like to use inline snapshots for simple text blocks like Paragraphs or Heading.

Reply:
I also like serialised content more. In playwright, the inline snapshots were amazing. I wish we could do this in Playwright as well. Makes it so much easier to read and immediately know what's going on. I don't like the `getBlocks()` much because it's very verbose, doesn't read as easily as HTML (HTML is more readable, which is why we have JSX). And when you make a bigger change affecting many tests, it's a joy to update all (inline) snapshots at once with the `-u` flag.

Reply:
IMO JSX is much more verbose than `getBlocks` though. With `getBlocks`, we know what we're testing and what we're expecting at first glance. While with `toMatchInlineSnapshot`, it's often required to read the comment or the test title above to fully understand it.

JSX is great for representing UI and is more readable than nested function declaration. However, we're not writing UI here, we have a specific part of the UI that we want to test against. I think they are different problems and thus have different best practices.

> And when you make a bigger change affecting many tests, it's a joy to update all (inline) snapshots at once with the `-u` flag.

While it's convenient, it's also pretty dangerous. If you make a change that affects many snapshots, then I think it's often a sign that we're duplicating our tests. Ideally, we should test the structure of the block only once, and write focused assertions on things that we care about in those tests.

Humans are bad at doing repetitive jobs. Reviewing a single snapshot might not be an issue, but reviewing dozens of similar snapshots is. We tend to skip through some of them thinking that they are fine, but errors might just happen to lie within them.

It's not that snapshots are bad, I agree it's a useful tool for testing the overall representation of a block, but we should use it sparingly, preferably only once. We already have a dedicated place to test all blocks and their serialized representation using snapshots, I don't see why we need to repeat them in our e2e tests too.

A real-life example is in the `font-size-picker` e2e test. We use 14 inline snapshots there, but almost all of them are difficult to understand without contexts like comments or test titles. As a reader, I have to jump around the code to understand what the test is about. The escape sequences like `\\"` for JSON stringified text don't help either.

---

With all that said, we don't have inline snapshots in Playwright, yet. Using regular snapshots will just create a bunch of files that make it even harder to review. I think `getBlocks` is a good enough alternative that encourages explicit and atomic assertions as a best practice. WDYT if we pursue this path forward?

Reply:
Not saying we should use JSX, I was talking about HTML and just making a comparison about readability.

If you pursue this, I'd love it if you try to make the assertion even shorter.

Why not just do `expect( await editor.getBlocks() ).toMatchObject( [] )` instead of checking the array length and then a separate expect call for every block?

Reply:
I think the same argument applies to HTML though. HTML is useful for representing UI, but not so much for isolating the focus of the test.

> Why not just do `expect( await editor.getBlocks() ).toMatchObject( [] )` instead of checking the array length and then a separate expect call for every block?

Yeah, that works too! I just want to show that we don't have to compare against all of the blocks. Instead, we should be encouraged to write assertions on only a small part of the blocks when needed. For instance, if we only care about the length of the blocks and the content of the first block, we can do so with two separate `expect`s rather than having to check the whole array. This is the whole point of this article: making atomic and explicit assertions.

However, I think it might be easier to follow in documentation with only a single `toMatchObject` though. Given that many people might just copy-paste code from there. I just updated the doc to do that 👍 .

Reply:
> I think the same argument applies to HTML though. HTML is useful for representing UI, but not so much for isolating the focus of the test.

HTML is much more readable. Imagine the HTML for a table block vs the block JSON object. Even for a simple paragraph I find it much more readable. But I don't mind, it's fine.

Reply:
I'm not denying HTML is readable, just that it might not be the best tool for the job. Imagine when the sole purpose of a test is to check for the content of a single column in the table block. Instead of comparing the full HTML of the table block, checking only the column makes the test easier to follow and review. It depends on the tests and their focus, but I often find comparing the full HTML in a snapshot out of focus. I rarely know what I should be looking for in those HTML when everything is equally important. 😅

Reply:
That's where we disagree I think. I'd agree when creating unit tests, but with e2e test I think it's good to check the whole content for any unintended side effects.

Reply:
I see! I guess my point is that we already have dedicated tests for that, and there are better ways than outputting the full HTML markup to whoever is reading the code.

May I suggest a middle ground of using the diff snapshot trick for this?

```js
// Insert the block...

const beforeContent = await editor.getEditedPostContent();

// Do something in the test...

const afterContent = await editor.getEditedPostContent();
expect( diff( beforeContent, afterContent ) ).toMatchInlineSnapshot();
```

I'm perfectly fine with this technique as it also encourages minimal atomic assertions. It's less readable without any comment than explicit assertions though, but easier to write. This way, we get to test the full content without having to sacrifice readability with long dreadful snapshots. The coverage for `beforeContent` should already be covered in "Full post content test fixtures" so we don't have to check for that again. Would this sound like a better alternative to you?

Reply:
There are a lot of diff functions out there that we can use. The most common one might be the one from Jest: [`jest-diff`](https://www.npmjs.com/package/jest-diff).

> I'm not talking about the initial content, but rather content after edits or other actions.

Yep, that's why I said we don't need to check the output of the initial content. We just want to check the "diff" between the initial one and the content after editing. I mentioned this just to note that we aren't sacrificing coverage for readability.